### ISSUE TITLE
A better way to deprecate 3DES and other ciphers with 64-bit blocks

### ISSUE THREAD

**Issue by zzq1015** [ssllabs/ssllabs-scan#463]
First, mark them as weak in the "Cipher Suite" section, and warn the users that they have 64-bit blocks;
Second, Only IE6 and IE8 on Windows XP does not support AES, and can only use 3DES, ANY other clients can do AES. So, cap the grade to C while 3DES is negotiated other than IE6-8 on WinXP;
Third, add `SWEET32 attack` underneath the `RC4` status in the "Protocol Details" section: show "yes" when negotiates 64-bit block ciphers with non-IE6-8 clients, show "undesirable" when those ciphers are put at the bottom, and show "no" when not supporting those ciphers.

**Comment by adamcaudill** [comment]
Keep in mind that negotiating 3DES does not indicate, by itself, that a system is vulnerable to SWEET32. SWEET32 requires a large amount of data be sent in a single session - this can be limited by a variety of factors, limiting Keep-Alive requests is the most common. If the amount of traffic is limited, then the system isn't vulnerable to SWEET32 even if 3DES is negotiated.

**Comment by zzq1015** [comment]
Well, in my opinion, once 3DES is enabled, the system is "vulnerable" (can be attacked theoretically), but not "exploitable" (can be attacked in practise) so far. However, attacks only gets better. And nowadays, using 3DES has absolutely no upsides except compatibility (<1/20 of the speed of AES, <128-bit security, no hardware acceleration like AES). Also, old clients like IE6-8 are really unusable on the modern Internet, so except for some large websites, along with some CDNs, there isn't any good reasons to support 3DES.
Most server admins knows nothing about server ciphers, and there are still not enough people even aware of the presence of SSL Labs, so anyone who actually test the servers are likely to be security-inclined. 
Security problems aside, back when RC4 was considered strong, RC4 was preferred over 3DES (and sometimes over AES!); now RC4 is totally broken, and most servers have hardware acceleration for AES, and quite a few even supports ChaCha20, 3DES isn't preferred at all except some careless server configurations, as older versions of OpenSSL uses AES256 > 3DES > AES128. 
And how about IDEA and Blowfish? AFAIK IDEA is not supported by any clients (browsers) except some SSL/TLS libraries, so basically nobody is using it. Blowfish is not even defined in the TLS standard.

**Comment by ArchangeGabriel** [comment]
@zzq1015 I agree with @adamcaudill that 3DES being broken depends on context. One major “pro” service provider where I live had no SSL/TLS on their mail servers for a long time, and less than one year ago, they started supporting it. This is their ciphers list (for TLS1.0/1/2, they don’t support SSLv2/3):
```
TLS_RSA_WITH_RC4_128_SHA
TLS_RSA_WITH_3DES_EDE_CBC_SHA
TLS_DHE_RSA_WITH_3DES_EDE_CBC_SHA
TLS_RSA_WITH_CAMELLIA_128_CBC_SHA
TLS_DHE_RSA_WITH_CAMELLIA_128_CBC_SHA
TLS_RSA_WITH_CAMELLIA_256_CBC_SHA
TLS_DHE_RSA_WITH_CAMELLIA_256_CBC_SHA
TLS_RSA_WITH_SEED_CBC_SHA
TLS_DHE_RSA_WITH_SEED_CBC_SHA
TLS_ECDHE_RSA_WITH_RC4_128_SHA
TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA
TLS_DH_anon_WITH_3DES_EDE_CBC_SHA
TLS_DH_anon_WITH_CAMELLIA_128_CBC_SHA
TLS_DH_anon_WITH_CAMELLIA_256_CBC_SHA
TLS_DH_anon_WITH_SEED_CBC_SHA
TLS_ECDH_anon_WITH_RC4_128_SHA
TLS_ECDH_anon_WITH_3DES_EDE_CBC_SHA
```
But indeed, this is for email servers. So for instance TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA is probably a good choice here, because 3DES is currently not broken in this context. Though I would agree moving from it would be a good idea, but regarding those servers that’s the last thing I’m concerned about. Note that I’ve questioned them about the lack of AES, but to no avail.

P.S.: In the meantime, I got one other major provider to upgrade from:
```
TLS_RSA_WITH_RC4_128_MD5
TLS_RSA_WITH_RC4_128_SHA
TLS_RSA_WITH_3DES_EDE_CBC_SHA
TLS_RSA_WITH_DES_CBC_SHA
TLS_RSA_EXPORT_WITH_RC4_40_MD5
TLS_RSA_EXPORT_WITH_RC2_CBC_40_MD5
```
to:
```
TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
TLS_DHE_RSA_WITH_AES_128_GCM_SHA256
TLS_DHE_RSA_WITH_AES_256_CBC_SHA256
TLS_DHE_RSA_WITH_AES_128_CBC_SHA256
TLS_RSA_WITH_AES_256_GCM_SHA384
TLS_RSA_WITH_AES_128_GCM_SHA256
TLS_RSA_WITH_AES_256_CBC_SHA256
TLS_RSA_WITH_AES_128_CBC_SHA256
TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
TLS_DHE_RSA_WITH_AES_256_CBC_SHA
TLS_DHE_RSA_WITH_AES_128_CBC_SHA
TLS_DHE_RSA_WITH_3DES_EDE_CBC_SHA
TLS_RSA_WITH_AES_256_CBC_SHA
TLS_RSA_WITH_AES_128_CBC_SHA
TLS_RSA_WITH_3DES_EDE_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA
```
This is current NIST/PCI-DSS/HIPAA recommendations. Note that listed 3DES cipher suites are the only way to contact previously mentioned provider, and I seem to recall they are the only mandatory cipher suites.

**Comment by zzq1015** [comment]
OK. Maybe it is still too early to retire 3DES completely. Keeping 3DES does not cause any harm for now as long as you keep it at the very bottom. However, so many servers uses `AES256 > 3DES > AES128`, or even worse, `(RC4 >)3DES > AES`. Maybe consider giving a harsher penalty to those servers.
@ArchangeGabriel in your first cipher list, there are so many bizarre problems:
1. Lack of [AES](https://tools.ietf.org/html/rfc3268), but with support of [ECDH](https://tools.ietf.org/html/rfc4492), [Camellia](https://tools.ietf.org/html/rfc4132), and [SEED](https://tools.ietf.org/html/rfc4162). Look at the publish date of each document.
1. The server supports \*\_anon\_\* ciphers, which is open to man-in-the-middle attack.
1. If the cipher list is what ciphers the server prefers, it has one more problem: RC4 preference, which is below today's security baseline.

The server admin probably weakened the configuration intentionally.

For the other server, I've got another problem: why does it prefer DHE instead of ECDHE? NIST curves are probably a concern, but why not use non-NIST curves like `Curve25519`? Preferring DHE slows down your server (longer handshake time), and very likely to cause compatibility problems with Java. [Java 6 and 7 supports up to 1024-bit DH parameters only](https://wiki.mozilla.org/Security/Server_Side_TLS#DHE_and_Java).

**Comment by ArchangeGabriel** [comment]
@zzq1015 Regarding your points:

1. Yeah, like I’ve said, I don’t understand why and they didn’t answer me on this. Maybe that’s because AES is NSA/NIST issued…
2. *_anon_* ciphers weren’t there when they initially turned TLS support on, so I would agree they intentionally weakened the server configuration, probably because some server somewhere only speaks those…
3. Not sure whether those ciphers are ordered by preference or if the server even provides a preference order.

For the other server, once again I’m not sure about order. Regarding NIST curves, I agree that it’s a problem, but the thing is `Curve25519` is not easily available right now (for instance, needs OpenSSL>1.1.x, which isn’t packaged by most Linux distros ATM). Regarding Java, I think it doesn’t matter here.

**Comment by bhushan5640** [comment]
All 64 bit block ciphers marked as "WEAK". 

Grades only crapped with 64 bit block ciphers negotiated with modern protocols (TLS 1.1 or greater)
https://blog.qualys.com/ssllabs/2017/01/18/ssl-labs-grading-changes-january-2017

Dev starts marking them weak
https://dev.ssllabs.com/ssltest/analyze.html?d=3des.badssl.com&hideResults=on

https://dev.ssllabs.com/ssltest/analyze.html?d=www.google.com&hideResults=on&latest
